### ISSUE TITLE
Gradients are not computed for input tensor if requires_grad=True

### ISSUE THREAD

**Issue by m1pro** [johnmarktaylor91/torchlens#34]
```python
 import torch
 from torch import nn
 import torchlens as tl

 torch.manual_seed(42)

 x = torch.randn(10, 8, requires_grad=True)
 model = nn.Linear(8, 1)
 model_history = tl.log_forward_pass(model, x, detach_saved_tensors=False, save_gradients=True)
 out = model_history[-1].tensor_contents
 out.backward(torch.randn_like(out))

 print(x.grad)


 # Expected:

 torch.manual_seed(42)
 x = torch.randn(10, 8, requires_grad=True)
 model = nn.Linear(8, 1)
 out = model(x)
 out.backward(torch.randn_like(out))

 print(x.grad)

```

output:

None
tensor([[-9.7358e-02, -1.2852e-01,  1.0003e-01, -7.6202e-02,  1.1442e-01,
         -4.4736e-02,  1.0784e-01, -1.4648e-01],
        [-5.0190e-01, -6.6254e-01,  5.1565e-01, -3.9284e-01,  5.8985e-01,
         -2.3063e-01,  5.5596e-01, -7.5511e-01],
        [ 7.0126e-02,  9.2571e-02, -7.2048e-02,  5.4888e-02, -8.2415e-02,
          3.2224e-02, -7.7679e-02,  1.0551e-01],
        [ 3.5828e-01,  4.7295e-01, -3.6810e-01,  2.8042e-01, -4.2106e-01,
          1.6463e-01, -3.9686e-01,  5.3903e-01],
        [ 5.6418e-02,  7.4475e-02, -5.7964e-02,  4.4158e-02, -6.6305e-02,
          2.5924e-02, -6.2494e-02,  8.4882e-02],
        [ 7.7931e-02,  1.0287e-01, -8.0066e-02,  6.0996e-02, -9.1587e-02,
          3.5810e-02, -8.6324e-02,  1.1725e-01],
        [-8.2235e-02, -1.0856e-01,  8.4489e-02, -6.4365e-02,  9.6646e-02,
         -3.7788e-02,  9.1092e-02, -1.2372e-01],
        [ 4.0560e-04,  5.3542e-04, -4.1672e-04,  3.1746e-04, -4.7668e-04,
          1.8638e-04, -4.4929e-04,  6.1023e-04],
        [ 3.6155e-02,  4.7726e-02, -3.7145e-02,  2.8298e-02, -4.2490e-02,
          1.6613e-02, -4.0048e-02,  5.4395e-02],
        [-3.2423e-01, -4.2800e-01,  3.3311e-01, -2.5377e-01,  3.8104e-01,
         -1.4898e-01,  3.5915e-01, -4.8780e-01]])

**Comment by johnmarktaylor91** [comment]
I went ahead and fixed this (pushed to GitHub and pip)—now it should save gradients for input and output tensors to the model as well. To access these, use `layer.grad_contents` for whatever layer you want (which can include input and output layers):

<img width="657" alt="Image" src="https://github.com/user-attachments/assets/a71187d9-9b49-4bd7-97c4-fcb03528470e" />

One thing to note: TorchLens deepcopies all input tensors prior to feeding them into the model, such that the original tensors remain unchanged. So in your example, `x.grad` will not contain the gradient, since it was deepcopied prior to the forward pass. But you can still access the input tensor gradient with `model_history[0].grad_contents`. Just curious, does this get in the way of your workflow in any way? If so I can think about adding an option to not deepcopy the original input tensors.

**Comment by m1pro** [comment]
For now  deep copy is fine. 
Thanks.

**Comment by m1pro** [comment]
I suggest adding the following test to your tests module:

```python
import torch
from torch import nn
import torchlens as tl


def test_input_grad_contents():

    torch.manual_seed(33)
    grad_output = torch.randn((10, 1))

    torch.manual_seed(42)
    x = torch.randn(10, 8, requires_grad=True)
    model = nn.Linear(8, 1)
    model_history = tl.log_forward_pass(model, x, detach_saved_tensors=False, save_gradients=True)
    out = model_history[-1].tensor_contents
    out.backward(grad_output)

    # Expected:
    torch.manual_seed(42)

    x = torch.randn(10, 8, requires_grad=True)
    model = nn.Linear(8, 1)
    out = model(x)
    out.backward(grad_output)

    assert torch.allclose(model_history[0].grad_contents, x.grad)
```
